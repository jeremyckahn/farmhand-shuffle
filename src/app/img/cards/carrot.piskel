{"modelVersion":2,"piskel":{"name":"carrot","description":"","fps":12,"height":24,"width":24,"layers":["{\"name\":\"Layer 1\",\"opacity\":1,\"frameCount\":1,\"chunks\":[{\"layout\":[[0]],\"base64PNG\":\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAA+klEQVRIS82UOw7CMBBEnQrRUHIBDpiKU1BxQC5ASYNSBdmJzXo/3jGSJVJa9jzPeDZTGPxNg/XDnwDuxzU5nd/dF8IPRIgGsNb37NsAepi7AF35DjgkugDFowkMkPPPMC0WA+oDojgV5o+dhY0S4AA6MDQmp10eYKvn5kLOJFDbFmB93Q5F9HRdahB3wR0CNa0Acb+AAMNnOSjij+cSLmfmhMfViMoFxEuqkFxxZyY0gHp7Ctmj+p5tQJoA7fbkLbwGpmfmm8zsq0dG/gBGi0RzfhXOraUOkriSNfbPkmMoIiq3L33viMLQl28ADKClpa5DTehSZJuHAz7EuYYZ/S2Q+AAAAABJRU5ErkJggg==\"}]}"]}}