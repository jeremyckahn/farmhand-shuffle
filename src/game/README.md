# Game logic

This directory contains Farmhand Shuffle game logic. Game logic is distinct from app logic. It is concerned with the mechanics of the core Farmhand Shuffle card game. It is not concerned with user experience or presentation.
