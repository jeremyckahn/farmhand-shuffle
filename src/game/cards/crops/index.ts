export * from './carrot'
export * from './pumpkin'
