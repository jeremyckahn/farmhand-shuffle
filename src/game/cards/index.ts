import { v4 as uuid } from 'uuid'

import { ICard, Instance } from '../types'

export * from './crops'
export * from './events'
export * from './water'

export const instantiate = <T = ICard>(card: T): T & Instance => {
  return { ...card, instanceId: uuid() }
}
