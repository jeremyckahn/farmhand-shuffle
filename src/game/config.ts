export const initialPlayerFunds = 50

export const initialHandSize = 7

export const deckSize = 60

export const standardTaxAmount = 5
