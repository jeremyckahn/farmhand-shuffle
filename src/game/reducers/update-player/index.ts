import { IGame, IPlayer } from '../../types'

export const updatePlayer = (
  game: IGame,
  playerId: IPlayer['id'],
  newPlayerProperties: Partial<IPlayer>
): IGame => {
  const newGame = structuredClone(game)

  newGame.table.players[playerId] = {
    ...game.table.players[playerId],
    ...newPlayerProperties,
  }

  return newGame
}
