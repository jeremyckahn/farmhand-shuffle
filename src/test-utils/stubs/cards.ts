import { carrot, instantiate, pumpkin, rain, water } from '../../game/cards'
import { CropInstance, EventInstance, WaterInstance } from '../../game/types'

export const stubCarrot: CropInstance = instantiate(carrot)
export const stubPumpkin: CropInstance = instantiate(pumpkin)
export const stubWater: WaterInstance = instantiate(water)
export const stubRain: EventInstance = instantiate(rain)
