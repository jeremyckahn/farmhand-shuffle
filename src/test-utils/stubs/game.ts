import { IGame } from '../../game/types'

import { stubTable } from './table'

export const stubGame = (): IGame => {
  return {
    table: stubTable(),
  }
}
