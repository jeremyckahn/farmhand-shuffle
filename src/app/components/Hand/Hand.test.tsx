import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { carrot, pumpkin, water } from '../../../game/cards/index'
import { updatePlayer } from '../../../game/reducers/update-player/index'
import { stubGame } from '../../../test-utils/stubs/game'

import {
  getGapPixelWidth,
  Hand,
  HandProps,
  selectedCardTransform,
} from './Hand'

const baseGame = stubGame()

const handCards = [carrot, pumpkin, water]
const game = updatePlayer(baseGame, baseGame.sessionOwnerPlayerId, {
  hand: handCards.map(({ id }) => id),
})

const StubHand = (overrides: Partial<HandProps>) => {
  return (
    <Hand game={game} playerId={game.sessionOwnerPlayerId} {...overrides} />
  )
}

describe('Hand', () => {
  test('renders hand with no cards are selected', () => {
    render(<StubHand />)

    for (const { name } of handCards) {
      const card = screen.getByText(name).closest('.MuiPaper-root')

      const { transform } = getComputedStyle(card!)
      expect(transform).toMatchSnapshot()
    }
  })

  test('renders hand with a card selected', async () => {
    render(<StubHand />)

    const card1 = screen.getByText(handCards[0].name).closest('.MuiPaper-root')
    await userEvent.click(card1!)

    const { transform: card1Tranform } = getComputedStyle(card1!)
    expect(card1Tranform).toEqual(selectedCardTransform)

    for (const { name } of handCards.slice(1)) {
      const card = screen.getByText(name).closest('.MuiPaper-root')

      const { transform } = getComputedStyle(card!)
      expect(transform).toMatchSnapshot()
    }
  })

  describe('getGapPixelWidth', () => {
    test.each([
      { numberOfCards: 0, gapSize: 50 },
      { numberOfCards: 6, gapSize: 30 },
      { numberOfCards: 11, gapSize: 15 },
      { numberOfCards: 21, gapSize: 10 },
      { numberOfCards: 31, gapSize: 5 },
      { numberOfCards: 61, gapSize: 3 },
      { numberOfCards: 1000, gapSize: 3 },
    ])(
      'calculates gap size for $numberOfCards cards',
      ({ gapSize, numberOfCards }) => {
        expect(getGapPixelWidth(numberOfCards)).toEqual(gapSize)
      }
    )
  })
})
