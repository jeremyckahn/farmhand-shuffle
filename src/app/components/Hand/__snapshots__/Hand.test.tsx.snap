// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hand renders hand with a card selected 1`] = `"translateX(calc(-50% + 50px + -50px)) translateY(calc(28rem / 2)) rotate(-5deg) scale(0.65) rotateY(25deg)"`;

exports[`Hand renders hand with a card selected 2`] = `"translateX(calc(-50% + 50px + 50px)) translateY(calc(28rem / 2)) rotate(-5deg) scale(0.65) rotateY(25deg)"`;

exports[`Hand renders hand with no cards are selected 1`] = `"translateX(calc(-50% + 50px + -150px)) translateY(0rem) rotate(-5deg) scale(1) rotateY(25deg)"`;

exports[`Hand renders hand with no cards are selected 2`] = `"translateX(calc(-50% + 50px + -50px)) translateY(0rem) rotate(-5deg) scale(1) rotateY(25deg)"`;

exports[`Hand renders hand with no cards are selected 3`] = `"translateX(calc(-50% + 50px + 50px)) translateY(0rem) rotate(-5deg) scale(1) rotateY(25deg)"`;
